File: VB/WindowsApplication123/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraBars
Imports DevExpress.XtraEditors.Repository

Namespace WindowsApplication123
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()

			mru = New MRUItemManager(barManager1, barLinkContainerItem1)
		End Sub

		Private mru As MRUItemManager

		Private Sub barButtonItem1_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barButtonItem1.ItemClick
		End Sub

		Private Sub barButtonItem2_ItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs) Handles barButtonItem2.ItemClick
			Dim res As DialogResult = openFileDialog1.ShowDialog()
			If res <> System.Windows.Forms.DialogResult.OK Then
				Return
			End If
			mru.Add(openFileDialog1.FileName)
		End Sub
	End Class

	Public Class MRUItemManager
		Private manager As BarManager
		Private list As List(Of BarEditItem)
		Private maxItems As Integer
		Private riCheck As RepositoryItemCheckEdit
		Private container As BarLinkContainerItem

		Public Sub New(ByVal manager As BarManager, ByVal container As BarLinkContainerItem)
			Me.manager = manager
			Me.list = New List(Of BarEditItem)()
			maxItems = 5
			riCheck = New RepositoryItemCheckEdit()
			riCheck.Caption = ""
			riCheck.GlyphAlignment = DevExpress.Utils.HorzAlignment.Far
			manager.RepositoryItems.Add(riCheck)
			Me.container = container
		End Sub

		Public Sub Add(ByVal obj As Object)
			If obj Is Nothing Then
				Return
			End If
			Dim item As BarEditItem = Nothing
			For i As Integer = 0 To list.Count - 1
				If Equals(list(i).Tag, obj) Then
					item = list(i)
					Exit For
				End If
			Next i
			If item Is Nothing Then
				item = New BarEditItem()
				manager.Items.Add(item)
				item.Edit = riCheck
				item.AutoFillWidth = True
				item.Caption = obj.ToString()
				item.Tag = obj
				item.EditValue = False
				AddHandler item.ItemClick, AddressOf item_ItemClick
				list.Add(item)
			End If
			Touch(item)
		End Sub

		Private Sub item_ItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
			e.Item.Manager.CloseMenus()
			Touch(TryCast(e.Item, BarEditItem))
		End Sub

		Protected Sub Touch(ByVal item As BarEditItem)
			list.Remove(item)
			list.Insert(0, item)
			Dim i As Integer = list.Count - 1
			Do While list.Count > maxItems AndAlso i >= 0
				Dim pin As Boolean = Equals(list(i).EditValue, True)
				If (Not pin) Then
					Dim xitem As BarEditItem = list(i)
					list.RemoveAt(i)
					xitem.Dispose()
				End If
				i -= 1
			Loop
			UpdateList()
		End Sub

		Public Sub UpdateList()
			container.ItemLinks.Clear()
			For i As Integer = 0 To list.Count - 1
				container.AddItem(list(i))
			Next i
		End Sub
	End Class
End Namespace

File: CS/WindowsApplication123/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraBars;
using DevExpress.XtraEditors.Repository;

namespace WindowsApplication123
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();

            mru = new MRUItemManager(barManager1, barLinkContainerItem1);
        }

        MRUItemManager mru;

        private void barButtonItem1_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
        }

        private void barButtonItem2_ItemClick(object sender, ItemClickEventArgs e)
        {
            DialogResult res = openFileDialog1.ShowDialog();
            if (res != DialogResult.OK) return;
            mru.Add(openFileDialog1.FileName);
        }
    }

    public class MRUItemManager
    {
        private BarManager manager;
        private List<BarEditItem> list;
        private int maxItems;
        private RepositoryItemCheckEdit riCheck;
        private BarLinkContainerItem container;

        public MRUItemManager(BarManager manager, BarLinkContainerItem container)
        {
            this.manager = manager;
            this.list = new List<BarEditItem>();
            maxItems = 5;
            riCheck = new RepositoryItemCheckEdit();
            riCheck.Caption = "";
            riCheck.GlyphAlignment = DevExpress.Utils.HorzAlignment.Far;
            manager.RepositoryItems.Add(riCheck);
            this.container = container;
        }

        public void Add(object obj)
        {
            if (obj == null) return;
            BarEditItem item = null;
            for (int i = 0; i < list.Count; i++)
            {
                if (Equals(list[i].Tag, obj))
                {
                    item = list[i];
                    break;
                }
            }
            if (item == null)
            {
                item = new BarEditItem();
                manager.Items.Add(item);
                item.Edit = riCheck;
                item.AutoFillWidth = true;
                item.Caption = obj.ToString();
                item.Tag = obj;
                item.EditValue = false;
                item.ItemClick += new ItemClickEventHandler(item_ItemClick);
                list.Add(item);
            }
            Touch(item);
        }

        void item_ItemClick(object sender, ItemClickEventArgs e)
        {
            e.Item.Manager.CloseMenus();
            Touch(e.Item as BarEditItem);
        }

        protected void Touch(BarEditItem item)
        {
            list.Remove(item);
            list.Insert(0, item);
            int i = list.Count - 1;
            while (list.Count > maxItems && i >= 0)
            {
                bool pin = Equals(list[i].EditValue, true);
                if (!pin)
                {
                    BarEditItem xitem = list[i];
                    list.RemoveAt(i);
                    xitem.Dispose();
                }
                i--;
            }
            UpdateList();
        }

        public void UpdateList()
        {
            container.ItemLinks.Clear();
            for (int i = 0; i < list.Count; i++)
            {
                container.AddItem(list[i]);
            }
        }
    }
}